drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua

```diff
@@ -0,0 +1,1580 @@
+  end
+end
+
+local function handle_unlock(driver, device, command)
+  local ep = device:component_to_endpoint(command.component)
+  local cota_cred = device:get_field(lock_utils.COTA_CRED)
+  if cota_cred then
+    device:send(
+      DoorLock.server.commands.UnlockDoor(device, ep, cota_cred)
+    )
+  else
+    device:send(DoorLock.server.commands.UnlockDoor(device, ep))
+  end
+end
+
+----------------
+-- User Table --
+----------------
+local function add_user_to_table(device, userIdx, usrType)
+  -- Get latest user table
+  local user_table = device:get_latest_state(
+    "main",
+    capabilities.lockUsers.ID,
+    capabilities.lockUsers.users.NAME
+  ) or {}
+  local new_user_table = {}
+
+  -- Recreate user table
+  for index, entry in pairs(user_table) do
+    table.insert(new_user_table, entry)
+  end
+
+  -- Add new entry to table
+  table.insert(new_user_table, {userIndex = userIdx, userType = usrType})
+  device:emit_event(capabilities.lockUsers.users(new_user_table, {visibility = {displayed = false}}))
+end
+
+local function update_user_in_table(device, userIdx, usrType)
+  -- Get latest user table
+  local user_table = device:get_latest_state(
+    "main",
+    capabilities.lockUsers.ID,
+    capabilities.lockUsers.users.NAME
+  ) or {}
+  local new_user_table = {}
+
+  -- Recreate user table
+  local i = 0
+  for index, entry in pairs(user_table) do
+    if entry.userIndex == userIdx then
+      i = index
+    end
+    table.insert(new_user_table, entry)
+  end
+
+  -- Update user entry
+  if i ~= 0 then
+    new_user_table[i].userType = usrType
+    device:emit_event(capabilities.lockUsers.users(new_user_table, {visibility = {displayed = false}}))
+  end
+end
+
+local function delete_user_from_table(device, userIdx)
+  -- If User Index is ALL_INDEX, remove all entry from the table
+  if userIdx == ALL_INDEX then
+    device:emit_event(capabilities.lockUsers.users({}, {visibility = {displayed = false}}))
+    return
+  end
+
+  -- Get latest user table
+  local user_table = device:get_latest_state(
+    "main",
+    capabilities.lockUsers.ID,
+    capabilities.lockUsers.users.NAME
+  ) or {}
+  local new_user_table = {}
+
+  -- Recreate user table
+  for index, entry in pairs(user_table) do
+    if entry.userIndex ~= userIdx then
+      table.insert(new_user_table, entry)
+    end
+  end
+  device:emit_event(capabilities.lockUsers.users(new_user_table, {visibility = {displayed = false}}))
+end
+
+----------------------
+-- Credential Table --
+----------------------
+local function add_credential_to_table(device, userIdx, credIdx, credType)
+  -- Get latest credential table
+  local cred_table = device:get_latest_state(
+    "main",
+    capabilities.lockCredentials.ID,
+    capabilities.lockCredentials.credentials.NAME
+  ) or {}
+  local new_cred_table = {}
+
+  -- Recreat credential table
+  for index, entry in pairs(cred_table) do
+    table.insert(new_cred_table, entry)
+  end
```

Comment:
Last one- why are we rebuilding the table? Thank you!

Reply:
I did not know how to use table api, so I implemented it as above.
If a new entry is added to the `cred_table` as an insert, there is an issue that the existing entries information disappears, so I made it as above. I implemented it as below, could you tell me which part is the problem?

```
local function add_credential_to_table(device, userIdx, credIdx, credType)
  -- Get latest credential table
  local cred_table = device:get_latest_state(
    "main",
    capabilities.lockCredentials.ID,
    capabilities.lockCredentials.credentials.NAME
  ) or {}

  -- Add new entry to table
  table.insert(cred_table, {userIndex = userIdx, credentialIndex = credIdx, credentialType = credType})
  device:emit_event(capabilities.lockCredentials.credentials(cred_table, {visibility = {displayed = false}}))
end
```

Reply:
This is true of all relevant credential functions. Though I don't know all of the details, the call to `device:get_latest_state` is pulling something out of memory that is not mutable in-place, which the function `table.insert()` attempts to do. Without getting into the specifics of _why_ this is the case, I'm gonna propose a couple alternatives.

First, I looked into the [lockCredentials capability definition](https://github.ecodesamsung.com/iot/SmartThingsCapabilities/blob/main/capabilities/lock-credentials.src/lock-credentials.yml) and it seems to supports a series of commands that directly handle this memory such as `addCredential`, `deleteCredential`, etc. I believe these could be leveraged to handle these systems.

In the case that these cannot be leveraged for whatever reason, I believe we can use the `deep_copy` function like so: `utils.deep_copy(device:get_latest_state())` which will do the same functionality as you describe here- creating a deep copy. This is currently used in the Aqara lock credential handler, see [here](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/main/drivers/Aqara/aqara-lock/src/credential_utils.lua#L49).

Again, I think the capability commands SHOULD do all of this for us, but in the case that it does not, we should leverage `deep_copy` here. Thanks!

cc: @ctowns so he knows this is still an open item

Reply:
Thank you for helping me~! I modified these parts related to the table api and it works well.